[SPARK-26030][BUILD] Bump previousSparkVersion in MimaBuild.scala to be 2.4.0

cloud-fan opened the pull request:
## What changes were proposed in this pull request?

Since Spark 2.4.0 is already in maven repo, we can Bump previousSparkVersion in MimaBuild.scala to be 2.4.0.

Note that, seems we forgot to do it for branch 2.4, so this PR also updates MimaExcludes.scala

## How was this patch tested?

N/A

SparkQA commented:
**[Test build #98596 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98596/testReport)** for PR 22977 at commit [`a035778`](https://github.com/apache/spark/commit/a0357782e75967d309dc5229b3c36a0c295f2956).

HyukjinKwon reviewed (approved):
Looks good if tests pass

cloud-fan commented:
I think one major issue is, there is no document about how to update mima with new releases. Anyone knows the detailed process? Seems we need to update `MimaExcludes.scala` with something like `lazy val v30excludes = v24excludes ++ Seq` when cut the new branch, and update `MimaBuild.scala` when the new release is published.

And there are 2 remaining issues.
1. the data source v2 changes broke a lot of mima rules, while I expect interfaces marked as "Envolving" should not be tracked by mima.
2. mllib broke a lot of mima rules, seems caused by https://github.com/apache/spark/pull/22921 . @srowen  can you take a look?

Also cc @JoshRosen @gatorsmile @shaneknapp @vanzin @holdenk @felixcheung

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/4848/
Test PASSed.

HyukjinKwon commented:
Actually, similar changes were being made at https://github.com/apache/spark/pull/22967 FYI. cc @dbtsai

SparkQA commented:
**[Test build #98596 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98596/testReport)** for PR 22977 at commit [`a035778`](https://github.com/apache/spark/commit/a0357782e75967d309dc5229b3c36a0c295f2956).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98596/
Test FAILed.

dongjoon-hyun commented on project/MimaBuild.scala:
```diff
@@ -88,7 +88,7 @@ object MimaBuild {
   def mimaSettings(sparkHome: File, projectRef: ProjectRef) = {
     val organization = "org.apache.spark"
-    val previousSparkVersion = "2.2.0"
+    val previousSparkVersion = "2.4.0"
```
According to https://issues.apache.org/jira/browse/SPARK-23070, `2.3.0` seems to be the correct version in this PR.

cc @gatorsmile .

cloud-fan replied:
up to my understanding, we should have changed it to 2.3.0 when 2.3.0 was released.

Maybe we should send a another PR to branch-2.4 and change it to 2.3.0

srowen replied:
I believe @cloud-fan is right here. If master is 3.0, the previous version is 2.4.0 in master. Analogously for branch 2.4. Yes sounds like this should be a step in the release process somewhere.

dongjoon-hyun replied:
Thanks, @cloud-fan and @srowen . +1 for another PR for `branch-2.4`.

cloud-fan commented on project/MimaExcludes.scala:
```diff
@@ -84,7 +84,17 @@ object MimaExcludes {
+    ProblemFilters.exclude[IncompatibleMethTypeProblem]("org.apache.spark.sql.expressions.Window.rangeBetween"),
+    ProblemFilters.exclude[IncompatibleMethTypeProblem]("org.apache.spark.sql.expressions.WindowSpec.rangeBetween"),
+    // [SPARK-23781][CORE] Merge token renewer functionality into HadoopDelegationTokenManager
+    ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.deploy.SparkHadoopUtil.nextCredentialRenewalTime"),
```
This is actually a private method, I'm not sure why mima tracks it.

srowen replied:
It was a `private[spark]` method which has to be implemented as a public method in the Java bytecode. MiMa doesn't know the difference. Yes it can be ignore for sure.

So these failures are 'new' because they were introduced after Spark 2.2 (hence not missing since 2.2) then removed, so bumping the previous version causes them to fail now? OK.

srowen commented on project/MimaExcludes.scala:
```diff
@@ -84,7 +84,17 @@ object MimaExcludes {
+    ProblemFilters.exclude[IncompatibleMethTypeProblem]("org.apache.spark.sql.expressions.WindowSpec.rangeBetween"),
+    // [SPARK-23781][CORE] Merge token renewer functionality into HadoopDelegationTokenManager
+    ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.deploy.SparkHadoopUtil.nextCredentialRenewalTime"),
+    // [SPARK-25908][CORE][SQL] Remove old deprecated items in Spark 3
```
@cloud-fan really these two lines should move to the top, because the first long chunk of exclusions is for 25908.

felixcheung commented:
I think also there is a hive metastore test that downloads spark release jar?

srowen commented:
HiveExternalCatalogVersionsSuite ? that's already updated.

felixcheung commented:
right, I mean both this and that should be part of the process "post-release"

cloud-fan committed:
[SPARK-26029][BUILD][2.4] Bump previousSparkVersion in MimaBuild.scala to be 2.3.0

## What changes were proposed in this pull request?

Although it's a little late, we should still update mima for branch 2.4, to avoid future breaking changes.

Note that, when merging, we should forward port it to master branch, so that the excluding rules are still in `v24excludes`.

TODO: update the release process document to mention about mima update.

## How was this patch tested?

N/A

Closes #23015 from cloud-fan/mima-2.4.

Authored-by: Wenchen Fan <wenchen@databricks.com>
Signed-off-by: Wenchen Fan <wenchen@databricks.com>

cloud-fan committed:
Bump previousSparkVersion in MimaBuild.scala to be 2.4.0

cloud-fan commented on project/MimaExcludes.scala:
```diff
@@ -164,7 +212,50 @@ object MimaExcludes {
-    ProblemFilters.exclude[MissingClassProblem]("org.apache.spark.ml.classification.LabelConverter")
+    ProblemFilters.exclude[MissingClassProblem]("org.apache.spark.ml.classification.LabelConverter"),
+
+    // [SPARK-21842][MESOS] Support Kerberos ticket renewal and creation in Mesos
```
these changes are cherry-picked from https://github.com/apache/spark/pull/23015

SparkQA commented:
**[Test build #98746 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98746/testReport)** for PR 22977 at commit [`8b9efe1`](https://github.com/apache/spark/commit/8b9efe14fa4c53fa2f13f598879d7e45c47d3a6c).

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
**[Test build #98750 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98750/testReport)** for PR 22977 at commit [`8d9f5c7`](https://github.com/apache/spark/commit/8d9f5c768415607b9aa779a6dee291724047d6b4).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98750/
Test FAILed.

SparkQA commented:
**[Test build #98753 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98753/testReport)** for PR 22977 at commit [`802b521`](https://github.com/apache/spark/commit/802b521989c4e4365dcc44df0bae4bcc505a7428).

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/4965/
Test PASSed.

dbtsai commented:
LGTM. Thanks!

cloud-fan commented:
since this PR only touches mima, and the jenkins already passed the mima check, I'm going to merge it to master, thanks!

SparkQA commented:
**[Test build #98753 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98753/testReport)** for PR 22977 at commit [`802b521`](https://github.com/apache/spark/commit/802b521989c4e4365dcc44df0bae4bcc505a7428).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98753/
Test FAILed.
